includes/panopto_derivatives.inc

```diff
@@ -4,38 +4,265 @@
+        'message_sub' => array(
+          '@response' => $response,
+          '@header' => $header,
+          '@content' => $content,
+        ),
+        'type' => 'watchdog',
+        'severity' => WATCHDOG_ERROR,
+      );
+      $failureMessage['messages'][] = $messages;
+    }
+  }
+  return $session;
+}
+
+/**
+ * Uploads file to Panopto using AWS S3 Uploader
+ * @param string $upload_target The URL of upload target
+ * @param string $file_path The URL which is the target of the request
+ * @throws Exception The S3Client may throw an exception which is unhandled by this method.
+ * @return void
+ */
+function islandora_video_streaming_panopto_upload_file($upload_target, $file_path) {
+  $element = explode("/", $upload_target);
+  $prefix = array_pop($element);
+  $service_endpoint = implode("/", $element);
+  $bucket = array_pop($element);
+  $file_name = basename($file_path);
+  $object_key = $prefix."/".$file_name;
+  
+  $s3Params = array(
+      'endpoint' => $service_endpoint,
+      'region'  => 'us-east-1',
+      'version' => '2006-03-01',
+      'credentials' => [
+        'key'    => 'dummy',
+        'secret' => 'dummy'
+      ]
+  );
+  // Create an S3Client
+  $s3Client = new S3Client($s3Params);
+
+  $source = fopen($file_path, 'rb');
+  $uploader = new ObjectUploader(
+    $s3Client,
+    $bucket,
+    $object_key,
+    $source
+  );
+
+  do {
+    try {
+      $result = $uploader->upload();
+    } catch (MultipartUploadException $e) {
+      rewind($source);
+      $uploader = new MultipartUploader($s3Client, $source, [
+        'state' => $e->getState(),
+      ]);
+    }
+  } while (!isset($result));
+}
+
+/**
+ * Creates manifest file, in temporary storage, needed for Panopto Video Upload
+ * @param string $file_path The filepath of the Islandora object
+ * @return string The filename or filepath of the newly created manifest file
+ */
+function islandora_video_streaming_panopto_create_manifest_for_video($file_path) {
+  $file_name = basename($file_path);
+  $manifest_template = drupal_get_path('module','islandora_video_streaming') . DIRECTORY_SEPARATOR . 'includes' . DIRECTORY_SEPARATOR . 'upload_manifest_template.xml';
+  $template = file_get_contents($manifest_template);
+  $template = str_replace("{Title}", $file_name, $template);
+  $template = str_replace("{Description}", 'This is a video session with the uploaded video file '.$file_name, $template);
+  $template = str_replace("{Filename}", $file_name, $template);
+  $template = str_replace("{Date}", date('c'), $template);
+  
+  $part_of_file_name = explode(".", $file_name);
+  $updated_manifest_file = tempnam('temporary://', $part_of_file_name[0]);
+  rename($updated_manifest_file, $updated_manifest_file .= '.xml');
+  file_put_contents($updated_manifest_file, $template);
+
+  return $updated_manifest_file;
+}
+
+/**
+ * Finishes upload
+ * @param string $baseUrl The base URL for request
```

Comment:
I do not know.

Reply:
An object-oriented approach would be helpful here.  For example, consider an object "PanoptoClient", which is constructed with a `baseUrl`, `clientId`, `clientSecret`, `user`, and `password`.  It could then hold protected properties of an established `$token` and `$session`, and support public methods like `uploadVideo()`.

This doesn't need to be done for this PR, but would make good future enhancement.